### DISCUSSION TITLE
Uploaded files keep show 'Waiting for size' and 'Loading' with spinning animation in filament admin for spatie media library field

### DISCUSSION THREAD

**Discussion by ziming** [filamentphp/filament#2696] [Help]
### Package

filament/spatie-laravel-media-library-plugin

### Package Version

2.12.27

### Laravel Version

9.17

### Livewire Version

2.10.5

### PHP Version

8.1.6

### Bug description

In my admin, in the edit and view view. the uploaded files look like this

https://share.getcloudapp.com/nOuXmkjq

You will spot a few things:

- file name is displayed correctly
- On the left side, It keep say Waiting for size forever. However we know that the spatie media library media table has a size column
- On the right side, it keep says "Loading" and 'tap to cancel'

Lastly, not sure if i should make a new discussion or if I can write it here.

It will be nice if on the right side of each file there is also a 'Download' circle icon in both View and Edit

**Comment by ziming** [comment]
Oh and if it helps, my files are on an s3 private bucket

My .env FILAMENT_FILESYSTEM_DRIVER did specify s3 too

**Comment by danharrin** [comment]
Any console errors?

**Reply by ziming** [reply]
Ah yes

```php
Access to fetch at 'https://my-sensitive-bucket-name.s3.amazonaws.com/uploads/6705/some-boring-report.pdf' from origin 'https://my-website.com' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
module.esm.js:21930          GET https://my-sensitive-bucket-name.s3.amazonaws.com/uploads/6705/some-boring-report.pdf net::ERR_FAILED 403
(anonymous) @ module.esm.js:21930
c @ runtime.js:63
(anonymous) @ runtime.js:294
(anonymous) @ runtime.js:119
Nl @ app.js?id=b9896eff36b8d71fcc4059e927e10783:8
a @ app.js?id=b9896eff36b8d71fcc4059e927e10783:8
(anonymous) @ app.js?id=b9896eff36b8d71fcc4059e927e10783:8
(anonymous) @ app.js?id=b9896eff36b8d71fcc4059e927e10783:8
(anonymous) @ module.esm.js:21929
n @ module.esm.js:13821
load @ module.esm.js:13808
l @ module.esm.js:14484
ADD_ITEM @ module.esm.js:15091
l @ module.esm.js:12226
(anonymous) @ module.esm.js:14766
DID_SET_FILES @ module.esm.js:14763
l @ module.esm.js:12226
r.<computed> @ module.esm.js:13234
l @ module.esm.js:12226
(anonymous) @ module.esm.js:15340
SET_OPTIONS @ module.esm.js:15339
l @ module.esm.js:12226
Zm @ module.esm.js:17872
Jm @ module.esm.js:18181
nv @ module.esm.js:18279
iv @ module.esm.js:18288
_v @ module.esm.js:18441
(anonymous) @ module.esm.js:21903
c @ runtime.js:63
(anonymous) @ runtime.js:294
(anonymous) @ runtime.js:119
Nl @ app.js?id=b9896eff36b8d71fcc4059e927e10783:8
a @ app.js?id=b9896eff36b8d71fcc4059e927e10783:8
Promise.then (async)
Nl @ app.js?id=b9896eff36b8d71fcc4059e927e10783:8
a @ app.js?id=b9896eff36b8d71fcc4059e927e10783:8
(anonymous) @ app.js?id=b9896eff36b8d71fcc4059e927e10783:8
(anonymous) @ app.js?id=b9896eff36b8d71fcc4059e927e10783:8
(anonymous) @ module.esm.js:21903
(anonymous) @ module.esm.js:457
W @ module.esm.js:434
(anonymous) @ module.esm.js:2622
(anonymous) @ module.esm.js:1152
n @ module.esm.js:546
(anonymous) @ module.esm.js:551
St @ module.esm.js:764
(anonymous) @ module.esm.js:724
start @ module.esm.js:723
884 @ app.js:66
i @ bootstrap:19
(anonymous) @ startup:4
i.O @ chunk loaded:23
(anonymous) @ startup:6
(anonymous) @ app.js?id=b9896eff36b8d71fcc4059e927e10783:66
Show 3 more frames
module.esm.js:21930                  Uncaught (in promise) TypeError: Failed to fetch
    at module.esm.js:21930:36
    at c (runtime.js:63:40)
    at Generator._invoke (runtime.js:294:22)
    at Generator.next (runtime.js:119:21)
    at Nl (app.js?id=b9896eff36b8d71fcc4059e927e10783:8:162572)
    at a (app.js?id=b9896eff36b8d71fcc4059e927e10783:8:162776)
    at app.js?id=b9896eff36b8d71fcc4059e927e10783:8:162837
    at new Promise (<anonymous>)
    at app.js?id=b9896eff36b8d71fcc4059e927e10783:8:162716
    at module.esm.js:21929:17

```

**Reply by ziming** [reply]
I see it says aws S3 cors error, anyone know what is the cors settings to set on aws s3 for a private bucket?

on another note, we actually don't need to fetch these metadata like size and such from the s3 bucket right? We can fetch those values from spatie media table and fill them in. Doing it this way will save the need to configure any cors setting on the aws s3 bucket.

I used to use plain filepond in the past and I gotten those values this way. Maybe the code can be edited in the package to do it this way.

```blade
<script>
FilePond.create(inputElement, {
                labelIdle: 'PDFs Only',
                acceptedFileTypes: 'application/pdf',
                allowRevert: true,
                files: [
            @foreach($user->getMedia('books') as $media)
            {
                source: '{{ $media->uuid }}',
                options: {
                    type: 'local',
                    file: {
                        name: '{{ $media->file_name }}',
                        size: '{{ $media->size }}',
                        type: '{{ $media->mime_type }}',
                    }
                }
            },
            @endforeach
            ],
                allowRemove: true,
                server: {
                    url: 'the-url-to-upload',
                    process: {
                        method: 'post',
                        headers: {
                            'Accept': 'application/json',
                            'X-Requested-With': 'XMLHttpRequest',
                            'X-CSRF-TOKEN': '{{ csrf_token() }}',
                        },
                    },
                    // source is media identifier (id or uuid). see the source in the for each loop earlier.
                    remove: (source, load, error) => {
                        fetch('the-url-to-delete' + source, { 
                          method: 'DELETE',
                          headers: {
                            'Accept': 'application/json',
                            'X-Requested-With': 'XMLHttpRequest',
                            'X-CSRF-TOKEN': '{{ csrf_token() }}',
                        },
                    })
                        error('Failed');
                        load();
                    },
                    revert: (uniqueFileId, load, error) => {
                        fetch('/url-to-revert/'+uniqueFileId+'/', {
                          method: 'DELETE', // i treated revert the same as delete
                          headers: {
                            'Accept': 'application/json',
                            'X-Requested-With': 'XMLHttpRequest',
                            'X-CSRF-TOKEN': '{{ csrf_token() }}',
                        },
                    })
                        error('Failed');
                        load();
                    }


                },


            });
</script>
```

**Comment by danharrin** [comment]
Also, you posted no code example.

**Reply by ziming** [reply]
I didn't post any code example because I felt there is nothing special but here you go

```php
SpatieMediaLibraryFileUpload::make('annual_report') // pdf file
                                    ->collection('annual-report'),

                             
                                SpatieMediaLibraryFileUpload::make('photos')
                                    ->multiple()
                                    ->collection('photos'),
```

**Reply by danharrin** [reply]
If it's a private bucket, dont you need `visibility('private')`? So we generate temporary URLs instead.

**Reply by ziming** [reply]
Hi @danharrin  thanks 4 the tip. i added visibility('private') after ur suggestion but the bug still happens.

**Reply by ziming** [reply]
I also added CORS setting to my S3 bucket. This is my configuration. The error still occur and dev tools console also still mention cors error

```json
[
    {
        "AllowedHeaders": [
            "Authorization"
        ],
        "AllowedMethods": [
            "GET",
            "HEAD"
        ],
        "AllowedOrigins": [
            "app.my-domain.com"
        ],
        "ExposeHeaders": [
            "Access-Control-Allow-Origin"
        ]
    }
]
```

**Reply by ziming** [reply]
Ok resolved my issue.

Turns out AWS documentation is wrong

https://aws.amazon.com/premiumsupport/knowledge-center/s3-configure-cors/

For AllowedOrigins, the domains listed there should be something like https://example.com and not just `example.com` as the aws documentation link shows

Having said that, this other approach below I mentioned will eliminate the need to have CORS configured on the s3 bucket and will work for more people.

https://github.com/laravel-filament/filament/discussions/2696#discussioncomment-2910653

**Comment by aura2020** [comment]
For those having the same issue on localhost , change APP_URL to localhost   and also include port number 
APP_URL=http://localhost:8000

**Comment by morpheus7CS** [comment]
Hey,

I had the same issue using Sail and had to enter into container shell and run `php artisan storage:link` in there. There was a warning that the storage link is already created, but it started working normally after that. 🤷‍♂️ 

Kind regards,
g

**Comment by horaceho** [comment]
In my case, it's caused by a CORS error:
> Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://xxxx-xxx-xxx-xxx-xxx.ngrok.io/storage/1000005/CdoFEeqweTX0lOLVmI7rtNId95WKLM-metaQmVndWlyYmV1XzUuanBlZw==-.jpg. (Reason: CORS header ‘Access-Control-Allow-Origin’ missing). Status code: 200.

![Screenshot 2022-09-17 at 3 21 31 PM](https://user-images.githubusercontent.com/6657/190845857-c7ef532c-c2fb-4700-8c86-72d87492359b.png)

**Reply by jareerzeenam** [reply]
Makesure you have `APP_URL=http://127.0.0.1:8000 `path in `.env` and check if the localhost url and your `app_url` match.

**Reply by JstnAxel** [reply]
You saved me, bro. Thanks! :)

**Reply by antoHero** [reply]
You are a lifesaver. Thanks

**Comment by stuartcusackie** [comment]
For Digital Ocean I had to add Cors policy to the bucket. Origins: *, Allowed Methods: GET.

Strange because the ralphjsmit media library package doesn't require this.
